### PR TITLE
 [camera] Implement {GetMin|GetMax|Set}Exposure

### PR DESCRIPTION
**Pull request by bbrto21**
This PR also includes: 
* [camera] Introduce CameraDeviceError
* [camera] Update minor things
* [camera] Fix a bug related to zoom
* [camera] Use LOG_WARN instead of LOG_ERROR

### CODE REVIEW

File: packages/camera/tizen/src/camera_device.cc [hunk 1]
```diff
@@ -935,48 +1022,55 @@ void CameraDevice::StartVideoRecording(
...
+    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>>
+        &&result) noexcept {
   LOG_DEBUG("enter");
   CommitRecorder();
   UnprepareRecorder();
 
-  StartCameraPreview();
+  if (StartCameraPreview()) {
+    std::string file_name;
+    GetRecorderFileName(file_name);
```
Comment (pwasowski2) [inline packages/camera/tizen/src/camera_device.cc:1043]: Shouldn't we check the error code here? If yes, please check the error code.

**Review by pwasowski2 — CHANGES_REQUESTED** [review]
Besides a few minor defects, LGTM

**Review by bwikbs — APPROVED** [review]
> I'm gonna to make a PR to merge this plugin to master after this PR

Thanks for working on the camera. I thought it would take a while, but It's not. :fireworks:

File: packages/camera/tizen/src/camera_device.cc [hunk 2]
```diff
@@ -404,12 +404,11 @@ bool CameraDevice::GetCameraZoomRange(int &min, int &max) {
   RETV_LOG_ERROR_IF(error != CAMERA_ERROR_NONE, false,
                     "camera_attr_get_zoom_range fail - error[%d]: %s", error,
                     get_error_message(error));
-  if (min > max) {
+  if (min >= max) {
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply packages/camera/tizen/src/camera_device.cc:407]: > > ```
> > If the min value is greater than the max value, it means that this feature is not supported. 
> > ```

You are right, but I found  a weird case with min and max are zero(front camera of TM1). In this case, the zoom behavior of the dart implementation is malfunctioning. I think that this is misleading, so I'll handle it as another if-statement

File: packages/camera/tizen/src/camera_device.cc [hunk 3]
```diff
@@ -733,39 +777,50 @@ bool CameraDevice::Open(std::string image_format_group) {
     SetCameraPreviewFormat(pixel_format);
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply packages/camera/tizen/src/camera_device.cc:777]: That's right, these things can fail depending on the device. But I did not check for errors intentionally.
This is to follow the original behavior and even if these two functions are failed, the Open must continue. 
And I think that there is nothing more to do at this time even if we check the return values(Each wrapper function logs an error)

File: packages/camera/tizen/src/camera_device.cc [hunk 4]
```diff
@@ -733,39 +777,50 @@ bool CameraDevice::Open(std::string image_format_group) {
...
-          surface);
-    }
+  if (!SetCameraMediaPacketPreviewCb([](media_packet_h pkt, void *data) {
+        tbm_surface_h surface = nullptr;
+        int error = media_packet_get_tbm_surface(pkt, &surface);
+        LOG_ERROR_IF(error != MEDIA_PACKET_ERROR_NONE,
+                     "media_packet_get_tbm_surface fail - error[%d]: %s", error,
+                     get_error_message(error));
+
+        if (error == 0) {
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply packages/camera/tizen/src/camera_device.cc:787]: the macro LOG_ERROR_IF is not have return statement

File: packages/camera/tizen/src/camera_device.cc [hunk 5]
```diff
@@ -733,39 +777,50 @@ bool CameraDevice::Open(std::string image_format_group) {
...
-    if (error == 0) {
-      CameraDevice *camera_device = (CameraDevice *)data;
-      FlutterMarkExternalTextureFrameAvailable(
-          camera_device->GetTextureRegistrar(), camera_device->GetTextureId(),
-          surface);
-    }
+  if (!SetCameraMediaPacketPreviewCb([](media_packet_h pkt, void *data) {
+        tbm_surface_h surface = nullptr;
+        int error = media_packet_get_tbm_surface(pkt, &surface);
+        LOG_ERROR_IF(error != MEDIA_PACKET_ERROR_NONE,
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply packages/camera/tizen/src/camera_device.cc:783]: the macro LOG_ERROR_IF is not have return statement

File: packages/camera/tizen/src/camera_device.cc [hunk 6]
```diff
@@ -733,39 +777,50 @@ bool CameraDevice::Open(std::string image_format_group) {
...
+                     "media_packet_get_tbm_surface fail - error[%d]: %s", error,
+                     get_error_message(error));
+
+        if (error == 0) {
+          CameraDevice *camera_device = (CameraDevice *)data;
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply packages/camera/tizen/src/camera_device.cc:788]: Thanks~

**Review by pwasowski2 — APPROVED** [review]
LGTM 👌

File: packages/camera/tizen/src/camera_device.cc [hunk 7]
```diff
@@ -465,6 +480,20 @@ bool CameraDevice::SetCameraExposureMode(CameraExposureMode mode) {
...
 }
 
+bool CameraDevice::GetCameraExposureRange(int &min, int &max) {
+  int error = camera_attr_get_exposure_range(camera_, &min, &max);
+  RETV_LOG_ERROR_IF(error != CAMERA_ERROR_NONE, false,
+                    " camera_attr_get_exposure_range fail - error[%d]: %s",
+                    error, get_error_message(error));
+  // According to the API doc, this means that it is not supported on device
+  if (min > max) {
+    // Not supported on TM1
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply packages/camera/tizen/src/camera_device.cc:490]: No, according to the API documentation, it is not dependent on device.
I just wanted to notice that it is not supported on TM1 that is currently the only one device can be tested

File: packages/camera/tizen/src/camera_device.cc [hunk 8]
```diff
@@ -733,39 +785,50 @@ bool CameraDevice::Open(std::string image_format_group) {
...
-  // Start or stop exposure accroding to ExpoureMode
-  SetExposureMode(exposure_mode_);
+  try {
+    // Start or stop focusing according to FocusMode
+    SetFocusMode(focus_mode_);
+    // Start or stop exposure accroding to ExpoureMode
+    SetExposureMode(exposure_mode_);
+  } catch (const CameraDeviceError &error) {
+    LOG_ERROR("[%s] %s", error.GetErrorCode().c_str(),
+              error.GetErrorMessage().c_str());
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply packages/camera/tizen/src/camera_device.cc:830]: You are right, I'll change this to ```LOG_WARN```

File: packages/camera/tizen/src/camera_device.cc [hunk 9]
```diff
@@ -819,68 +885,92 @@ void CameraDevice::RestFocusPoint() {
...
 void CameraDevice::ResumeVideoRecording(
-    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>> &&result) {
+    std::unique_ptr<flutter::MethodResult<flutter::EncodableValue>>
+        &&result) noexcept {
   LOG_DEBUG("enter");
-  StartRecorder();
+  if (StartRecorder()) {
+    result->Success();
+  } else {
+    result->Error(kCameraDeviceError, "Failed to resume recorder");
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply packages/camera/tizen/src/camera_device.cc:894]: I want to call UpdateStates for both cases.

File: packages/camera/tizen/src/camera_device.cc [hunk 10]
```diff
@@ -935,48 +1030,60 @@ void CameraDevice::StartVideoRecording(
           ? locked_orientation_
           : orientation_manager_->GetDeviceOrientationType()));
   PrepareRecorder();
-  StartRecorder();
 
+  if (StartRecorder()) {
+    result->Success();
+  } else {
+    result->Error(kCameraDeviceError, "Failed to start recorder");
```
[the remark this answers is not in the document]
Comment (bbrto21) [reply packages/camera/tizen/src/camera_device.cc:1037]: I want to call UpdateStates for both cases.

**Review by pkosko — APPROVED** [review]
👍

**Pull request merged**